lib/client/ipam_test.go

```diff
@@ -105,6 +105,68 @@ type testArgsClaimAff struct {
 
 var _ = testutils.E2eDatastoreDescribe("IPAM tests", testutils.DatastoreEtcdV2, func(config api.CalicoAPIConfig) {
 
+	FDescribe("IPAM ReleaseIPs with duplicate IP can result in assigning an IP multiple times", func() {
+		c := testutils.CreateCleanClient(config)
+		ic := setupIPAMClient(c, true)
+
+		// The IP which we will cause a multiple assignment of
+		assignedIP := net.ParseIP("10.0.0.1")
+		pool := cnet.MustParseNetwork("10.0.0.0/24")
+		testutils.CreateNewIPPool(*c, "10.0.0.0/24", false, false, true)
+
+		Context("Setup duplicate oridinal in Unallocated list", func() {
+			// Assign the target IP
+			assignIPutil(ic, assignedIP, "host-a")
+
+			// Assign the rest of the IPs in the pool
+			ips, _, bulkAssignErr := ic.AutoAssign(client.AutoAssignArgs{
+				Num4:      255,
+				Hostname:  "host-1",
+				IPv4Pools: []cnet.IPNet{pool},
+			})
+
+			It("Should allocate most of the pool", func() {
+				Expect(bulkAssignErr).NotTo(HaveOccurred())
+				Expect(ips).NotTo(ContainElement(assignedIP))
+				Expect(len(ips)).To(Equal(255))
+			})
+
+			_, releaseErr := ic.ReleaseIPs([]cnet.IP{
+				cnet.IP{assignedIP},
+				cnet.IP{assignedIP},
```

Comment:
This is the essence of the issue - ReleaseIPs does not sanitize the IP list given nor does AllocationBlock's release() method.